File: packages/python/src/axonpush/eval/reports.py
"""Artifacts a CI run leaves behind: JSON, JUnit XML and a GitHub summary.

Byte-comparable with the TypeScript and .NET CLIs' artifacts.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from pathlib import Path
from xml.sax.saxutils import escape

TESTSUITE_NAME = "axonpush.evaluation"


@dataclass
class ItemResult:
    """One dataset example, as the local evaluator answered it."""

    item_id: str
    status: str
    output: object | None = None
    error: str | None = None
    latency_ms: float | None = None
    total_tokens: int | None = None
    cost_usd: float | None = None
    trace_id: str | None = None


@dataclass
class GateVerdict:
    passed: bool
    reasons: list[str] = field(default_factory=list)
    metrics: dict[str, object] = field(default_factory=dict)
    gate_run_id: str | None = None


@dataclass
class GitLineage:
    git_commit: str | None = None
    git_branch: str | None = None
    git_dirty: bool | None = None


@dataclass
class RunResult:
    experiment_id: str
    dataset_id: str
    dataset_revision: str
    started_at: str = ""
    completed_at: str = ""
    results: list[ItemResult] = field(default_factory=list)
    gate: GateVerdict | None = None
    cancelled: bool = False
    lineage: GitLineage = field(default_factory=GitLineage)


def _camel(name: str) -> str:
    head, *rest = name.split("_")
    return head + "".join(part.title() for part in rest)


def _camel_keys(value: object) -> object:
    if isinstance(value, dict):
        return {_camel(str(key)): _camel_keys(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_camel_keys(item) for item in value]
    return value


def to_json_report(run: RunResult) -> str:
    payload = {
        "experimentId": run.experiment_id,
        "datasetId": run.dataset_id,
        "datasetRevision": run.dataset_revision,
        "startedAt": run.started_at,
        "completedAt": run.completed_at,
        "cancelled": run.cancelled,
        "lineage": _camel_keys(
            {key: value for key, value in asdict(run.lineage).items() if value is not None}
        ),
        "results": [_camel_keys(asdict(item)) for item in run.results],
        "gate": _camel_keys(asdict(run.gate)) if run.gate else None,
    }
    return json.dumps(payload, indent=2, default=str) + "\n"


def _xml(value: str) -> str:
    return escape(value, {'"': "&quot;", "'": "&apos;"})


def to_junit_xml(run: RunResult) -> str:
    blocked = bool(run.gate and not run.gate.passed)
    failures = sum(1 for item in run.results if item.status != "passed") + int(blocked)
    skipped = sum(1 for item in run.results if item.status == "cancelled")
    lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        f'<testsuite name="{TESTSUITE_NAME}" tests="{len(run.results) + int(blocked)}" '
        f'failures="{failures}" skipped="{skipped}" timestamp="{_xml(run.started_at)}">',
    ]
    for item in run.results:
        attrs = (
            f'classname="{TESTSUITE_NAME}" name="{_xml(item.item_id)}" '
            f'time="{(item.latency_ms or 0) / 1000:.3f}"'
        )
        if item.status == "passed":
            lines.append(f"  <testcase {attrs}/>")
            continue
        kind = "skipped" if item.status == "cancelled" else "failure"
        lines.append(
            f'  <testcase {attrs}><{kind} message="{_xml(item.error or item.status)}"/></testcase>'
        )
    if run.gate and not run.gate.passed:
        lines.append(
            f'  <testcase classname="{TESTSUITE_NAME}" name="release gate" time="0.000">'
            f'<failure message="gate failed">{_xml("; ".join(run.gate.reasons))}</failure>'
            "</testcase>"
        )
    lines.append("</testsuite>")
    return "\n".join(lines) + "\n"


def _status_line(run: RunResult, failed: int) -> str:
    if run.cancelled:
        return "Cancelled"
    if run.gate:
        return "Passed" if run.gate.passed else "Failed"
    return "Completed with failures" if failed else "Passed"


def to_github_summary(run: RunResult) -> str:
    passed = sum(1 for item in run.results if item.status == "passed")
    failed = sum(1 for item in run.results if item.status == "failed")
    commit = run.lineage.git_commit
    lines = [
        "## AxonPush evaluation",
        "",
        f"**{_status_line(run, failed)}** · {passed}/{len(run.results)} item(s) passed"
        + (f" · {failed} failed" if failed else ""),
        "",
        "| Experiment | Dataset revision | Commit |",
        "| --- | --- | --- |",
        f"| `{run.experiment_id}` | `{run.dataset_id}@{run.dataset_revision}` | "
        + (f"`{commit[:12]}`" if commit else "—")
        + " |",
    ]
    if run.gate:
        lines += ["", "### Release gate", ""]
        lines.append("Gate passed." if run.gate.passed else "Gate failed.")
        if run.gate.reasons:
            lines += [""] + [f"- {reason}" for reason in run.gate.reasons]
    unsuccessful = [item for item in run.results if item.status != "passed"]
    if unsuccessful:
        lines += ["", "### Failed items", ""]
        lines += [f"- `{item.item_id}`: {item.error or item.status}" for item in unsuccessful]
    return "\n".join(lines) + "\n"


def write_artifact(path: str | None, contents: str) -> None:
    """Write an artifact, creating the directory. A missing path is a no-op."""
    if not path:
        return
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(contents, encoding="utf-8")

File: packages/dotnet/src/AxonPush.Cli/Reports.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AxonPush.Cli;

internal sealed class ItemResult
{
    [JsonPropertyName("itemId")] public string ItemId { get; init; } = string.Empty;
    [JsonPropertyName("status")] public string Status { get; init; } = "failed";
    [JsonPropertyName("output")] public object? Output { get; init; }
    [JsonPropertyName("error")] public string? Error { get; init; }
    [JsonPropertyName("latencyMs")] public double LatencyMs { get; init; }
    [JsonPropertyName("totalTokens")] public double? TotalTokens { get; init; }
    [JsonPropertyName("costUsd")] public double? CostUsd { get; init; }
    [JsonPropertyName("traceId")] public string? TraceId { get; init; }
}

internal sealed class GateVerdict
{
    [JsonPropertyName("passed")] public bool Passed { get; init; }
    [JsonPropertyName("reasons")] public IReadOnlyList<string> Reasons { get; init; } = [];
    [JsonPropertyName("metrics")]
    public IReadOnlyDictionary<string, object?> Metrics { get; init; }
        = new Dictionary<string, object?>();
    [JsonPropertyName("gateRunId")] public string? GateRunId { get; init; }
}

internal sealed class GitLineage
{
    [JsonPropertyName("gitCommit")] public string? GitCommit { get; init; }
    [JsonPropertyName("gitBranch")] public string? GitBranch { get; init; }
    [JsonPropertyName("gitDirty")] public bool? GitDirty { get; init; }
}

internal sealed class RunResult
{
    [JsonPropertyName("experimentId")] public string ExperimentId { get; init; } = string.Empty;
    [JsonPropertyName("datasetId")] public string DatasetId { get; init; } = string.Empty;
    [JsonPropertyName("datasetRevision")] public string DatasetRevision { get; init; } = string.Empty;
    [JsonPropertyName("startedAt")] public string StartedAt { get; init; } = string.Empty;
    [JsonPropertyName("completedAt")] public string CompletedAt { get; set; } = string.Empty;
    [JsonPropertyName("cancelled")] public bool Cancelled { get; set; }
    [JsonPropertyName("lineage")] public GitLineage Lineage { get; init; } = new();
    [JsonPropertyName("results")] public IReadOnlyList<ItemResult> Results { get; set; } = [];
    [JsonPropertyName("gate")] public GateVerdict? Gate { get; set; }
}

/// <summary>
/// Artifacts a CI run leaves behind. Byte-comparable with the TypeScript and
/// Python CLIs, because a pipeline is written against the report, not the
/// language that produced it.
/// </summary>
internal static class Reports
{
    private const string SuiteName = "axonpush.evaluation";

    private static readonly JsonSerializerOptions Json = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    public static string ToJsonReport(RunResult run) =>
        JsonSerializer.Serialize(run, Json) + "\n";

    private static string Xml(string value) => value
        .Replace("&", "&amp;", StringComparison.Ordinal)
        .Replace("<", "&lt;", StringComparison.Ordinal)
        .Replace(">", "&gt;", StringComparison.Ordinal)
        .Replace("\"", "&quot;", StringComparison.Ordinal)
        .Replace("'", "&apos;", StringComparison.Ordinal);

    public static string ToJUnitXml(RunResult run)
    {
        var blocked = run.Gate is { Passed: false };
        var failures = run.Results.Count(item => item.Status != "passed") + (blocked ? 1 : 0);
        var skipped = run.Results.Count(item => item.Status == "cancelled");
        var builder = new StringBuilder();
        builder.Append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
        builder.Append(CultureInfo.InvariantCulture, $"<testsuite name=\"{SuiteName}\" tests=\"{run.Results.Count + (blocked ? 1 : 0)}\" failures=\"{failures}\" skipped=\"{skipped}\" timestamp=\"{Xml(run.StartedAt)}\">\n");

        foreach (var item in run.Results)
        {
            var attrs = string.Create(
                CultureInfo.InvariantCulture,
                $"classname=\"{SuiteName}\" name=\"{Xml(item.ItemId)}\" time=\"{item.LatencyMs / 1000:F3}\"");
            if (item.Status == "passed")
            {
                builder.Append(CultureInfo.InvariantCulture, $"  <testcase {attrs}/>\n");
                continue;
            }

            var kind = item.Status == "cancelled" ? "skipped" : "failure";
            builder.Append(CultureInfo.InvariantCulture,
                $"  <testcase {attrs}><{kind} message=\"{Xml(item.Error ?? item.Status)}\"/></testcase>\n");
        }

        if (run.Gate is { Passed: false } gate)
        {
            // The gate is the point of the run, so CI should show it as a failing
            // test rather than only as a non-zero exit code.
            builder.Append(CultureInfo.InvariantCulture,
                $"  <testcase classname=\"{SuiteName}\" name=\"release gate\" time=\"0.000\"><failure message=\"gate failed\">{Xml(string.Join("; ", gate.Reasons))}</failure></testcase>\n");
        }

        builder.Append("</testsuite>\n");
        return builder.ToString();
    }

    public static string ToGitHubSummary(RunResult run)
    {
        var passed = run.Results.Count(item => item.Status == "passed");
        var failed = run.Results.Count(item => item.Status == "failed");
        var status = run.Cancelled
            ? "Cancelled"
            : run.Gate is { } verdict
                ? verdict.Passed ? "Passed" : "Failed"
                : failed > 0 ? "Completed with failures" : "Passed";
        var commit = run.Lineage.GitCommit;
        var lines = new List<string>
        {
            "## AxonPush evaluation",
            "",
            $"**{status}** · {passed}/{run.Results.Count} item(s) passed"
                + (failed > 0 ? $" · {failed} failed" : string.Empty),
            "",
            "| Experiment | Dataset revision | Commit |",
            "| --- | --- | --- |",
            $"| `{run.ExperimentId}` | `{run.DatasetId}@{run.DatasetRevision}` | "
                + (string.IsNullOrEmpty(commit) ? "—" : $"`{commit[..Math.Min(12, commit.Length)]}`")
                + " |",
        };

        if (run.Gate is { } gate)
        {
            lines.AddRange(["", "### Release gate", "", gate.Passed ? "Gate passed." : "Gate failed."]);
            if (gate.Reasons.Count > 0)
            {
                lines.Add("");
                lines.AddRange(gate.Reasons.Select(reason => $"- {reason}"));
            }
        }

        var unsuccessful = run.Results.Where(item => item.Status != "passed").ToList();
        if (unsuccessful.Count > 0)
        {
            lines.AddRange(["", "### Failed items", ""]);
            lines.AddRange(unsuccessful.Select(item => $"- `{item.ItemId}`: {item.Error ?? item.Status}"));
        }

        return string.Join("\n", lines) + "\n";
    }

    /// <summary>Write an artifact, creating the directory. A missing path is a no-op.</summary>
    public static async Task WriteArtifactAsync(string? path, string contents)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            return;
        }

        var directory = Path.GetDirectoryName(Path.GetFullPath(path));
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }

        await File.WriteAllTextAsync(path, contents).ConfigureAwait(false);
    }
}
